test/main-test.js

```diff
@@ -464,5 +464,150 @@ vows.describe('Regexp Tokenizer')
+          topic: ret('[\\-\\--\\-]'),
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  { type: types.CHAR, value: 45 },
+                  { type: types.RANGE, from: 45, to: 45 },
+                ],
+              }],
+            });
+          },
+        },
+        'token.from is a hyphen and the range is preceded by a predefined set [\\w\\--\\-]': {
+          topic: ret('[\\w\\--\\-]'),
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  {
+                    type: types.SET, not: false, set: [
+                      { type: types.CHAR, value: 95 },
+                      { type: types.RANGE, from: 97, to: 122 },
+                      { type: types.RANGE, from: 65, to: 90 },
+                      { type: types.RANGE, from: 48, to: 57 },
+                    ],
+                  },
+                  { type: types.RANGE, from: 45, to: 45 },
+                ],
+              }],
+            });
+          },
+        },
+        'token.from is a caret and the range is the first item of the set [\\^-9]': {
+          topic: ret('[\\^-9]'),
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  { type: types.RANGE, from: 94, to: 57 },
+                ],
+              }],
+            });
+          },
+        },
+        'token.to is a closing square bracket [2-\\]]': {
+          topic: ret('[2-\\]]'),
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  { type: types.RANGE, from: 50, to: 93 },
+                ],
+              }],
+            });
+          },
+        },
+        'token.to is a closing square bracket [\\^-\\]]': {
+          topic: ret('[\\^-\\]]'),
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  { type: types.RANGE, from: 94, to: 93 },
+                ],
+              }],
+            });
+          },
+        },
+        'token.to is a closing square bracket [[-\\]]': {
+          topic: ret('[[-\\]]'),
+          'Tokenizes correctly': t => {
+            assert.deepStrictEqual(t, {
+              type: types.ROOT, stack: [{
+                type: types.SET, not: false, set: [
+                  { type: types.RANGE, from: 91, to: 93 },
+                ],
+              }],
+            });
+          },
+        },
```

Comment:
maybe add closing square bracket that isn't escaped, as that is valid too. `[[-]]`